Add NCCL Alltoall to PT NCCL process group

srinivas212 opened the pull request:
Summary: Add Alltoall and Alltoallv to PT NCCL process group using NCCL Send/Recv.

Differential Revision: D22676566

fbshipit-source-id: f5efedc60f705d7b2612d53f981f955a95493025

Fixes #{issue number}

rohan-varma committed:
[Resubmit #41318] NCCL backend support for torch bool (#41959)

Summary:
Resubmit of https://github.com/pytorch/pytorch/issues/41318 pushed to ci-all branch.

Original description:
Closes https://github.com/pytorch/pytorch/issues/24137.
This PR adds support for the torch.bool tensor type to ProcessGroupNCCL. For most types we use the existing mapping, but since bool is not supported as a native ncclDataType_t, we add the following logic:

Map at::kBool to ncclUint8
During reduction (allreduce for example), if the operation is SUM, we instead override to to a MAX, to avoid overflow issues. The rest of the operations work with no changes. In the boolean case, changing sum to max makes no correctness difference since they both function as a bitwise OR.
The reduction logic (for example for reduce/allreduce) is as follows:
sum, max = bitwise or
product, min = bitwise and

Note that this PR doesn't add support for BAND/BOR/BXOR. That is because these reduction ops currently are not supported by NCCL backend, see https://github.com/pytorch/pytorch/issues/41362

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41959

Reviewed By: mrshenli

Differential Revision: D22719665

Pulled By: rohan-varma

fbshipit-source-id: 8bc4194a8d1268589640242277124f277d2ec9f1

Commit:
Updates torch.tensor, torch.as_tensor, and sparse ctors to use the device of inputs tensors they're given, by default (#41984)

Summary:
**BC-Breaking Note**

This PR changes the behavior of the torch.tensor, torch.as_tensor, and sparse constructors. When given a tensor as input and a device is not explicitly specified, these constructors now always infer their device from the tensor. Historically, if the optional dtype kwarg was provided then these constructors would not infer their device from tensor inputs. Additionally, for the sparse ctor a runtime error is now thrown if the indices and values tensors are on different devices and the device kwarg is not specified.

**PR Summary**
This PR's functional change is a single line:

```
auto device = device_opt.has_value() ? *device_opt : (type_inference ? var.device() : at::Device(computeDeviceType(dispatch_key)));
```
=>
```
auto device = device_opt.has_value() ? *device_opt : var.device();
```

in `internal_new_from_data`. This line entangled whether the function was performing type inference with whether it inferred its device from an input tensor, and in practice meant that

```
t = torch.tensor((1, 2, 3), device='cuda')
torch.tensor(t, dtype=torch.float64)
```

would return a tensor on the CPU, not the default CUDA device, while

```
t = torch.tensor((1, 2, 3), device='cuda')
torch.tensor(t)
```

would return a tensor on the device of `t`!

This behavior is niche and odd, but came up while aocsa was fixing https://github.com/pytorch/pytorch/issues/40648.

An additional side affect of this change is that the indices and values tensors given to a sparse constructor must be on the same device, or the sparse ctor must specify the dtype kwarg. The tests in test_sparse.py have been updated to reflect this behavior.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41984

Reviewed By: ngimel

Differential Revision: D22721426

Pulled By: mruberry

fbshipit-source-id: 909645124837fcdf3d339d7db539367209eccd48

Commit:
Move qconfig removal into convert() (#41930)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41930

As title
ghstack-source-id: 108517079

Test Plan: CI

Reviewed By: jerryzh168

Differential Revision: D22698386

fbshipit-source-id: 4f748c9bae4a0b615aa69c7cc8d8e451e5d26863

Commit:
update cub submodule (#42042)

Summary:
Per title

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42042

Reviewed By: mruberry

Differential Revision: D22752345

Pulled By: ngimel

fbshipit-source-id: 363735bfe3d49bab12fedef43b68c9dc9e372815

mrshenli committed:
Revert D22741514: [pytorch][PR] Enable ProcessGroupGlooTest in CI

Test Plan: revert-hammer

Differential Revision:
D22741514 (https://github.com/pytorch/pytorch/commit/45e6f2d60012df37393f5c5153e805209713ef19)

Original commit changeset: 738d2e27f523

fbshipit-source-id: 0381105ed0ab676b0abd1927f602a35b1b264a6a

mattip committed:
DOC: fix two build warnings (#41334)

Summary:
xref gh-38011.

Fixes two warnings when building documentation by
- using the external link to torchvision
- install tensorboard before building documentation

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41334

Reviewed By: ngimel

Differential Revision: D22753083

Pulled By: mruberry

fbshipit-source-id: 876377e9bd09750437fbfab0378664b85701f827

mattip committed:
DOC: split quantization.rst into smaller pieces (#41321)

Summary:
xref gh-38010 and gh-38011.

After this PR, there should be only two warnings:
```
pytorch/docs/source/index.rst:65: WARNING: toctree contains reference to nonexisting \
      document 'torchvision/index'
WARNING: autodoc: failed to import class 'tensorboard.writer.SummaryWriter' from module \
     'torch.utils'; the following exception was raised:
No module named 'tensorboard'
```

If tensorboard and torchvision are prerequisites to building docs, they should be added to the `requirements.txt`.

As for breaking up quantization into smaller pieces: I split out the list of supported operations and the list of modules to separate documents. I think this makes the page flow better, makes it much "lighter" in terms of page cost, and also removes some warnings since the same class names appear in multiple sub-modules.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41321

Reviewed By: ngimel

Differential Revision: D22753099

Pulled By: mruberry

fbshipit-source-id: d504787fcf1104a0b6e3d1c12747ec53450841da

rutujak24 committed:
Grammar Changes (#42076)

Summary:
Small grammatical updates.
![Screenshot (188)](https://user-images.githubusercontent.com/56619747/88471271-02723480-cf25-11ea-8fd1-ae98d5ebcc86.png)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42076

Reviewed By: mrshenli

Differential Revision: D22756651

Pulled By: ngimel

fbshipit-source-id: e810eb7397a5831d801348c8fff072854658830e

malfet committed:
Enable ProcessGroupGlooTest in CI (take 2) (#42086)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/42073

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42086

Reviewed By: ngimel

Differential Revision: D22765777

Pulled By: malfet

fbshipit-source-id: ebbcd44f448a1e7f9a3d18fa9967461129dd1dcd

alexandrosstergiou committed:
Added support for Huber Loss (#37599)

Summary:
Current losses in PyTorch only include a (partial) implementation of Huber loss through `smooth l1` based on Fast RCNN - which essentially uses a delta value of 1. Changing/Renaming the [`_smooth_l1_loss()`](https://github.com/pytorch/pytorch/blob/3e1859959a3f720bb5f5e47c3ca15fb3cbfae4da/torch/nn/functional.py#L2487) and refactoring to include delta, enables to use the actual function.

Supplementary to this, I have also made a functional and criterion versions for anyone that wants to set the delta explicitly - based on the functional `smooth_l1_loss()` and the criterion `Smooth_L1_Loss()`

Pull Request resolved: https://github.com/pytorch/pytorch/pull/37599

Differential Revision: D21559311

Pulled By: vincentqb

fbshipit-source-id: 34b2a5a237462e119920d6f55ba5ab9b8e086a8c

venkatacrc committed:
Remove settings for the logit test case. (#42114)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42114

Remove settings for the logit test case.

(Note: this ignores all push blocking failures!)

Test Plan: test_op_nnpi_fp16.py test case.

Reviewed By: hyuen

Differential Revision: D22766728

fbshipit-source-id: 2fe8404b103c613524cf1beddf1a0eb9068caf8a

alvgaona committed:
Amend docstring and add test for Flatten module (#42084)

Summary:
I've noticed when PR https://github.com/pytorch/pytorch/issues/22245 introduced `nn.Flatten`, the docstring had a bug where it wouldn't render properly on the web, and this PR addresses that. Additionally, it adds a unit test for this module.

**Actual**
![image](https://user-images.githubusercontent.com/13088001/88483672-cf896a00-cf3f-11ea-8b1b-a30d152e1368.png)

**Expected**
![image](https://user-images.githubusercontent.com/13088001/88483642-86391a80-cf3f-11ea-8333-0964a027a172.png)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42084

Reviewed By: mrshenli

Differential Revision: D22756662

Pulled By: ngimel

fbshipit-source-id: 60c58c18c9a68854533196ed6b9e9fb0d4f83520

Commit:
Updates to Scale and Zero Point Gradient Calculation (#42034)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42034

In this diff, scale and zero point gradient calculations are updated to correctly reflect the actual backpropagation equation (instead of `dScale * dX`, the near-final output should be `dScale * dY`; the same applies to zero point).

Test Plan:
To execute the unit tests for all affected learnable fake quantize modules and kernels, on a devvm, execute the following command:

`buck test //caffe2/test:quantization -- learnable`

To enable the `cuda` tests, execute the following command:

`buck test mode/dev-nosan //caffe2/test:quantization -- learnable`

Reviewed By: jerryzh168

Differential Revision: D22735668

fbshipit-source-id: 45c1e0fd38cbb2d8d5e60be4711e1e989e9743b4

Commit:
Updates to Python Module for Calculation of dX and Addition of Unit Tests (#42033)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42033

In this diff, the Python `_LearnableFakeQuantize` module is updated where the gradient with respect to the input `x` is actually computed instead of passed through. Argument naming is also updated for better clarity; and unit tests on the `PerTensor` and `PerChannel` operators are added for asserting correctness.

Test Plan:
On a devvm, execute the command:

`buck test //caffe2/test:quantization -- learnable_py_module`

To include `cuda` tests as well, run:

`buck test mode/dev-nosan //caffe2/test:quantization -- learnable_py_module`

Reviewed By: jerryzh168

Differential Revision: D22735580

fbshipit-source-id: 66bea7e9f8cb6422936e653500f917aa597c86de

Commit:
Renaming variables from dX to dY in Learnable Fake Quantize kernels for Better Clarity (#42032)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42032

In this diff, the arguments `dX` within the C++ kernels are named as `dY` for clarity and avoid confusion since it doesn't represent the gradient with respect to the input.

Test Plan:
To test all related fake quantize kernel operators, on a devvm, run the command:

`buck test //caffe2/test:quantization -- learnable`

Reviewed By: z-a-f, jerryzh168

Differential Revision: D22735429

fbshipit-source-id: 9d6d967f08b98a720eca39a4d2280ca8109dcdd6

smessmer committed:
Avoid refcount bump in IValue::toStringRef() (#42019)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42019

According to benchmarks, this makes IValue::toStringRef() 3-4x as fast.
ghstack-source-id: 108451154

Test Plan: unit tests

Reviewed By: ezyang

Differential Revision: D22731354

fbshipit-source-id: 3ca3822ea7310d8593e38b1d3e6014d6d80963db

IvanKobzarev committed:
[vulkan] Shaders caching (#39384)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39384

Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders),
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`

Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Test Plan: Imported from OSS

Reviewed By: AshkanAliabadi

Differential Revision: D21962430

Pulled By: IvanKobzarev

fbshipit-source-id: effe60538308805f317c11448b31dbcf670487e8

mariosasko committed:
Raise error for duplicate params in param group #40967 (#41597)

Summary:
This PR fixes an issue in https://github.com/pytorch/pytorch/issues/40967 where duplicate parameters across different parameter groups are not allowed, but duplicates inside the same parameter group are accepted. After this PR, both cases are treated equally and raise `ValueError`.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41597

Reviewed By: zou3519

Differential Revision: D22608019

Pulled By: vincentqb

fbshipit-source-id: 6df41dac62b80db042cfefa6e53fb021b49f4399

lly-zero-one committed:
Add a new op for converting the dense feature to sparse representation

Summary: we need this op to avoid the splicing of a dense tensor and then use the Mergesinglescaler op

Test Plan: integrated test with dper2

Differential Revision: D22677523

fbshipit-source-id: f4f9a1f06841b0906ec8cbb435482ae0a89e1721

acxz committed:
Find hip properly (#42064)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41886

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42064

Reviewed By: seemethere

Differential Revision: D22757115

Pulled By: malfet

fbshipit-source-id: 9c8805e6eb0b7d7defe0ecb08c1e45dcc775a237

Commit:
register parameters correctly in c++ MultiheadAttention (#42037)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42037

This is to fix #41951

Test Plan: Imported from OSS

Reviewed By: yf225

Differential Revision: D22764717

Pulled By: glaringlee

fbshipit-source-id: e6da0aeb05a2356f52446e6d5fad391f2cd1cf6f

subedigaurav committed:
change 2 instances of blacklist to blocklist in tools/pyi/gen_pyi.py (#41979)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41722

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41979

Reviewed By: ngimel

Differential Revision: D22764112

Pulled By: zou3519

fbshipit-source-id: 3f8580c96cf45078a9df3cd9ca6fdb10d58e143f

mrshenli commented:
I am not sure if this would trigger all ~100 master CI tests. I have only tried creating a PR from ci-all branch to master, like #42128. We will find out soon.

seemethere committed:
.circleci: Prefer netrc for docs push (#42136)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42136

Expect was giving weird issues so let's just use netrc since it doesn't
rely on janky expect behavior

Another follow up for: https://github.com/pytorch/pytorch/pull/41964

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Test Plan: Imported from OSS

Reviewed By: yns88

Differential Revision: D22778940

Pulled By: seemethere

fbshipit-source-id: 1bdf879a5cfbf68a7d2d34b6966c20f95bd0a3b5

ann-ss committed:
Refactor lite serializer dependencies from full jit (#42127)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42127

This diff renames core_autograd_sources to core_trainer_sources and moves/adds dependencies for the lite trainer in order to build the serializer functionality internally.
ghstack-source-id: 108589416

Test Plan: Manually tested serializer functionality from the internal lite trainer and verified that data is written correctly.

Reviewed By: iseeyuan

Differential Revision: D22738293

fbshipit-source-id: 992beb0c4368b2395f5bd5563fb2bc12ddde39a1

Commit:
Enable non-synchronizing cub scan for cum* operations (#42036)

Summary:
This uses cub for cum* operations, because, unlike thrust, cub is non-synchronizing.
Cub does not support more than `2**31` element tensors out of the box (in fact, due to cub bugs the cutoff point is even smaller)
so to support that I split the tensor into `2**30` element chunks, and modify the first value of the second and subsequent chunks to contain the cumsum result of the previous chunks. Since modification is done inplace on the source tensor, if something goes wrong and we error out before the source tensor is reverted back to its original state, source tensor will be corrupted, but in most cases errors will invalidate the full coda context.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42036

Reviewed By: ajtulloch

Differential Revision: D22749945

Pulled By: ngimel

fbshipit-source-id: 9fc9b54d466df9c8885e79c4f4f8af81e3f224ef

admud committed:
replace blacklist in aten/src/ATen/native/cudnn/Conv.cpp (#41627)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41700.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41627

Reviewed By: gmagogsfm

Differential Revision: D22678492

Pulled By: SplitInfinity

fbshipit-source-id: 75b82bd10059754d8e6c25fc20e9dde775d54698

agniutkarsh committed:
changed white-allowlisted (#41796)

Summary:
closes https://github.com/pytorch/pytorch/issues/41749

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41796

Reviewed By: gmagogsfm

Differential Revision: D22718991

Pulled By: SplitInfinity

fbshipit-source-id: 6c2d2b0e3b1e79fd515f9bdd395335a32f525a26

vybv committed:
Back out "Support aarch32 neon backend for Vec256"

Summary: Original commit changeset: 1c22cf67ec35

Test Plan: sandcastle, testing on Portal

Reviewed By: currybeef

Differential Revision: D22774614

fbshipit-source-id: 8897aec5df32092c4df86c0d54b0d2fe58d66e66

wconstab committed:
Add suggestion to enumerate ModuleDict in error message (#41946)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41946

Reviewed By: ngimel

Differential Revision: D22774243

Pulled By: wconstab

fbshipit-source-id: 5cfbe52b5b1c540f824593e67ae6ba4973458bb5

izmailovpavel committed:
Documentation for `torch.optim.swa_utils` (#41228)

Summary:
This PR adds a description of `torch.optim.swa_utils` added in https://github.com/pytorch/pytorch/pull/35032 to the docs at `docs/source/optim.rst`. Please let me know what you think!

vincentqb andrewgordonwilson

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41228

Reviewed By: ngimel

Differential Revision: D22609451

Pulled By: vincentqb

fbshipit-source-id: 8dd98102c865ae4a074a601b047072de8cc5a5e3

mrshenli commented:
Looks like this does not trigger all master test. Could you please try push a commit into `ci-all/ssrinivas` and then create a PR from that branch to master?

mrshenli committed:
Enable c10d Store tests in CI (#42128)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42128

Reviewed By: pritamdamania87

Differential Revision: D22774445

Pulled By: mrshenli

fbshipit-source-id: 6e5e56f42833414ef375b6cd23fdb3260cb07be9

xcheng16 committed:
[ModelLints] Refine dropout lint message. (#42046)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42046

Refine dropout lint message as we have enabled dropout operator removal in optimize_for_mobile method.
ghstack-source-id: 108607182

Test Plan: buck test ai_infra/ai_mobile_infra/tests:mobile_model_util_tests

Reviewed By: kimishpatel

Differential Revision: D22741132

fbshipit-source-id: 8f87356aae2bd9c89d1cad0d7be7286278bb14ad

mrshenli committed:
Let DDP.train() return self to stay consistent with nn.Module (#42131)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42131

Test Plan: Imported from OSS

Reviewed By: pritamdamania87

Differential Revision: D22775311

Pulled By: mrshenli

fbshipit-source-id: ac9e6cf8b2381036a2b6064bd029dca361a81777

tracelogfb committed:
onnxifi: make it work with AsyncIf

Summary:
the onnxifi path didn't handle the input/output name rewrite for ssa correctly for AsyncIf op. Add support for it.

Also fixed a place where we lose the net type while doing onnxifi transform.

Test Plan: Load 163357582_593 which is a multi feed model that uses AsyncIf. This used to fail with c2 not finding some blobs in workspace. Now it works.

Reviewed By: dhe95

Differential Revision: D21268230

fbshipit-source-id: ce7ec0e952513d0f251df1bfcfb2b0250f51fd94

chocjy committed:
Log the net if blob doesn't exist when setting output record (#41971)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41971

Reviewed By: wx1988

Differential Revision: D22490309

fbshipit-source-id: d967ee211b610f5523a307b5266b9fcb0277a21c

scintiller committed:
Make `run_binary_ops_test` function generic and Add tests to add_kernel function (#42101)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42101

1. Add test fixture `atest class` to store global variables
2. Make `run_binary_ops_test` function generic: can dispose different dtypes and different numbers of parameters
3. add test to `add_kernel`

Test Plan:
Run locally to check cover the corresponding code part in `BinaryOpsKernel.cpp`.
CI

Reviewed By: malfet

Differential Revision: D22760015

fbshipit-source-id: 95b47732f661124615c0856efa827445dd714125

Commit:
If we don't collect tracing, always free the trace data (#42118)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42118

We toggle trace on with a certain probablility. In the case of 3 inferences with trace on/off/on. We leak the trace from the first inference. Always clean up the trace will fix it.

Test Plan:
predictor

I created a tiny repro here: D22786551

With this fix, this issue is gone.

Reviewed By: gcatron

Differential Revision: D22768382

fbshipit-source-id: 9ee0bbcb2bc5f76107dae385759fe578909a683d

IvanKobzarev committed:
[vulkan] adaptive_avg_pool2d (#41220)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41220

Test Plan: Imported from OSS

Reviewed By: AshkanAliabadi

Differential Revision: D22754943

Pulled By: IvanKobzarev

fbshipit-source-id: 91a94f32db005ebb693384f4d27efe66e2c33a14

Commit:
[BugFix] Fix bug in onnx::SsaRewrite (#42148)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42148

Differential Revision: D22687388

fbshipit-source-id: facf7a186dd48d6f919d0ff5d42f756977c3f9f4

Commit:
[caffe2][tpx] Use logger instead of print

Test Plan: CI?

Differential Revision: D22790238

fbshipit-source-id: c0a801cdf7f0da489c67708a0eb1b498ff104c64

Commit:
Revert D22790238: [caffe2][tpx] Use logger instead of print

Test Plan: revert-hammer

Differential Revision:
D22790238 (https://github.com/pytorch/pytorch/commit/3c6fae6567230da0e08788b968cfb466ec629d65)

Original commit changeset: c0a801cdf7f0

fbshipit-source-id: cadfbd22f7d3ce656624483c9a19062f7c9a5b61

malfet committed:
Use new sccache for RocM builds (#42134)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42134

Reviewed By: seemethere

Differential Revision: D22782146

Pulled By: malfet

fbshipit-source-id: 85ba69a705600e30ae0eddbf654298b3dc6f96ed

zou3519 committed:
Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)` (#41942)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41942

This function:
- permutes all batch dims to the front of the tensors
- aligns all the batch dims to the collective levels of all the tensors
- expands all of the batch dims such that they are present in each of
the result tensors

This function is useful for the next diff up on the stack (which is
implementing a fallback kernel for BatchedTensor). It's also useful in
general for implementing batching rules on operators that take in
multiple batch dimensions at the front of each tensor (but we don't have
too many of those in PyTorch).

Test Plan: - `./build/bin/vmap_test`

Reviewed By: ezyang

Differential Revision: D22764104

Pulled By: zou3519

fbshipit-source-id: d42cc8824a1bcf258687de164b7853af52852f53

zou3519 committed:
Optimize alignBatchDimsAtFront (#41941)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41941

If we know that the tensor already has the desired aligned size, we
don't need to put in the effort to align it.

Test Plan: - `./build/bin/vmap_test`, `pytest test/test_vmap.py -v`

Reviewed By: albanD

Differential Revision: D22764101

Pulled By: zou3519

fbshipit-source-id: a2ab7ce7b98d405ae905f7fd98db097210bfad65

zou3519 committed:
Fix const correctness for VmapPhysicalView struct methods (#41940)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41940

See title. I marked methods that don't mutate the VmapPhysicalView as
`const`.

Test Plan: - wait for tests

Reviewed By: albanD

Differential Revision: D22764102

Pulled By: zou3519

fbshipit-source-id: 40f957ad61c85f0e5684357562a541a2712b1f38

Commit:
[caffe2][tpx] log to stderr (#42162)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42162

Test Plan: CI

Reviewed By: mrshenli

Differential Revision: D22791440

fbshipit-source-id: 14f16cd7a94a57161c5724177b518527f486232d

Commit:
Replace blacklist with blocklist in test/run_test.py file. (#42011)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41716
test/run_test.py file updated with an appropriate replacement for blacklist and whitelist.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42011

Reviewed By: pbelevich

Differential Revision: D22791836

Pulled By: malfet

fbshipit-source-id: 8139649c5b70c876b711e25c33f3051ea8461063

kalmufti committed:
Replace whitelist with allowlist (#42067)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41757

I've replaced all the whitelist with allowlist for this issue.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42067

Reviewed By: pbelevich

Differential Revision: D22791690

Pulled By: malfet

fbshipit-source-id: 638c13cf49915f5c83bd79c7f4a39b8390cc15b4

mattip committed:
replace onnx producer_version when updating results (#41910)

Summary:
xref gh-39002 which handled the reading but not the writing of the onnx expect files, and the last comment in that PR which points out `XXX` was suboptimal.
xref [this comment](https://github.com/pytorch/pytorch/pull/37091#discussion_r456460168) which pointed out the problem.

This PR:
- replaces `XXX` with `CURRENT_VERSION` in the stored files
- ensures that updating the results with the `--accept` flag will maintain the change

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41910

Reviewed By: pbelevich

Differential Revision: D22758671

Pulled By: ezyang

fbshipit-source-id: 47c345c66740edfc8f0fb9ff358047a41e19b554

Flamefire committed:
Don't run tests with custom arguments with pytest (#41397)

Summary:
This patch basically removes the `-m pytest` parameters when `extra_unittest_args` is used (e.g. `--subprocess`)

Fixes https://github.com/pytorch/pytorch/issues/41393

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41397

Reviewed By: pbelevich

Differential Revision: D22792133

Pulled By: ezyang

fbshipit-source-id: 29930d703666f4ecc0d727356bbab4a5f7ed4860

leaTepes committed:
Remove py2 specific code from test_utils.py (#42105)

Summary:
As https://github.com/pytorch/pytorch/issues/23795 mentioned drop Python 2 support. albanD
Fixes https://github.com/pytorch/pytorch/issues/31796

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42105

Reviewed By: ngimel

Differential Revision: D22765768

Pulled By: mrshenli

fbshipit-source-id: bae114a21cd5598004c7f92d313938ad826b4a24

daikikatsuragawa committed:
Document formatting (#42065)

Summary:
Apply syntax highlighting to the command in `README.md`. This makes `README.md` easier to read.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42065

Reviewed By: pbelevich

Differential Revision: D22753418

Pulled By: mrshenli

fbshipit-source-id: ebfa90fdf60478c34bc8a7284d163e0254cfbe3b

chengjinfang1216 committed:
Fix the issue GPU skip message(#41378) (#41973)

Summary:
Related https://github.com/pytorch/pytorch/issues/41378

Fix the issue GPU skip message

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41973

Reviewed By: pbelevich

Differential Revision: D22753459

Pulled By: mrshenli

fbshipit-source-id: d24b531926e28b860ae90b9ae07e8ca3438d21db

YifanShenSZ committed:
Avoid zero division in _cubic_interpolate (#42093)

Summary:
I encountered a zero division problem when using LBFGS:

File "/home/yshen/anaconda3/lib/python3.7/site-packages/torch/optim/lbfgs.py", line 118, in _strong_wolfe
    bracket[1], bracket_f[1], bracket_gtd[1])
File "/home/yshen/anaconda3/lib/python3.7/site-packages/torch/optim/lbfgs.py", line 21, in _cubic_interpolate
    d1 = g1 + g2 - 3 * (f1 - f2) / (x1 - x2)
ZeroDivisionError: float division by zero

My solution is to determine whether "line-search bracket is so small" before calling _cubic_interpolate

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42093

Reviewed By: pbelevich

Differential Revision: D22770667

Pulled By: mrshenli

fbshipit-source-id: f8fdfcbd3fd530235901d255208fef8005bf898c

xuhdev committed:
Make fmod work with zero divisors consistently (#41948)

Summary:
Currently `torch.tensor(1, dtype=torch.int).fmod(0)` crashes (floating point exception).

This PR should fix this issue.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41948

Reviewed By: ngimel

Differential Revision: D22771081

Pulled By: ezyang

fbshipit-source-id: a94dd35d6cd85daa2d51cae8362004e31f97989e

jspark1105 committed:
Back out "[NCCL] DDP communication hook: getFuture()" (#42152)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42152

Original commit changeset: 8c059745261d

Test Plan: .

Reviewed By: ajtulloch, jianyuh

Differential Revision: D22786183

fbshipit-source-id: 51155389d37dc82ccb4d2fa20d350f9d14abeaca

malfet committed:
Skip DataIO tests relying on LevelDB if compiled without it (#42169)

Summary:
Found while trying to get RocM Caffe2 job green

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42169

Reviewed By: seemethere

Differential Revision: D22791896

Pulled By: malfet

fbshipit-source-id: 9df6233876aec5ead056365499bab970aa7e8bdc

priyanshuwustl committed:
replace black list with block (#42091)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41729

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42091

Reviewed By: pbelevich

Differential Revision: D22792096

Pulled By: ezyang

fbshipit-source-id: caafa42d12cbad377b67ddbaba8f84a2b8c98066

xwang233 committed:
Raise RuntimeError for zero stride pooling (#41819)

Summary:
Close https://github.com/pytorch/pytorch/issues/41767

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41819

Reviewed By: mrshenli

Differential Revision: D22780634

Pulled By: ngimel

fbshipit-source-id: 376ce5229ad5bd60804d839340d2c6505cf3288d

IvanKobzarev committed:
[vulkan] mm op through addmm (#41221)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41221

Test Plan: Imported from OSS

Reviewed By: AshkanAliabadi

Differential Revision: D22754938

Pulled By: IvanKobzarev

fbshipit-source-id: f9a0f48d7943a85b7dbb3fc9edf9e214ba07543b

IvanKobzarev committed:
[vulkan] support add for dim < 4 (#41222)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41222

Test Plan: Imported from OSS

Reviewed By: AshkanAliabadi

Differential Revision: D22754937

Pulled By: IvanKobzarev

fbshipit-source-id: f8c5e55c965c0a805e75c63b21f410fb0c323515

malfet committed:
Skip TestExtractPredictorNet if compiled without OpenCV (#42168)

Summary:
Found while trying to get RocM Caffe2 CI green

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42168

Reviewed By: seemethere

Differential Revision: D22791879

Pulled By: malfet

fbshipit-source-id: 8f7ef9711bdc5941b2836e4c8943bb95c72ef8af

rohan-varma committed:
Allow drop_last option in DistributedSampler (#41171)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41171

DistributedSampler allows data to be split evenly across workers in
DDP, but it has always added additional samples in order for the data to be
evenly split in the case that the # of samples is not evenly divisible by the
number of workers. This can cause issues such as when doing distributed
validation accuracy, where multiple samples could be considered twice.

This PR adds a drop_last option where the tail of the data is dropped such that
the effective dataset size is still evenly divisible across the workers. This
ensures that DDP can train fine (there is no uneven inputs) and each replica
gets an equal number of data indices.
ghstack-source-id: 108617516

Test Plan: Added unittest

Reviewed By: mrshenli

Differential Revision: D22449974

fbshipit-source-id: e3156b751f5262cc66437b9191818b78aee8ddea

nitish-awasthi committed:
Replaced blacklist with blocklist (#42097)

Summary:
Closes https://github.com/pytorch/pytorch/issues/41726

Fixes https://github.com/pytorch/pytorch/issues/41726

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42097

Reviewed By: ngimel

Differential Revision: D22779535

Pulled By: SplitInfinity

fbshipit-source-id: 1d414af22a1b3e856a11d64cff4b4d33160d957b

kalmufti committed:
Replaced whitelist reference with allowlist (#42071)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41741

Replaced whitelist reference with allowlist.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42071

Reviewed By: pbelevich

Differential Revision: D22795176

Pulled By: SplitInfinity

fbshipit-source-id: bcf1b8afe516b9684ce0298bc257ef81152ba20c

nitish-awasthi committed:
Changed Blacklisted to Blocklisted (#42100)

Summary:
Fixes https://github.com/pytorch/pytorch/issues/41703

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42100

Reviewed By: ngimel

Differential Revision: D22780380

Pulled By: SplitInfinity

fbshipit-source-id: d465c41f1d4951ab6de55cb827c7ef53975209af

malfet committed:
Enable caffe2 tests for RocM jobs (#41604)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/41604

Reviewed By: ezyang

Differential Revision: D22603703

Pulled By: malfet

fbshipit-source-id: 789ccf2bb79668a5a68006bb877b2d88fb569809

ljk53 committed:
[pytorch] fix code analyzer for LLVM 9 & 10 (#42135)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42135

Tested the code analyzer with LLVM 9 & 10 and fixed a couple issues:
- Rename local demangle() which is available as public API since LLVM 9;
- Fix falsely associated op registrations due to the `phi` instruction;

Test Plan: Imported from OSS

Reviewed By: iseeyuan

Differential Revision: D22795508

Pulled By: ljk53

fbshipit-source-id: 2d47af088acd3312a7ea5fd9361cdccd48940fe6

wconstab committed:
make FusionCallback refer to specified GraphFuser context (#41560)

Summary:
Fixes issue where
 - top level fuser's block_ was captured by callback due to [&] capture,
 - recursive/nested fusers would compare erroneously to top-level block_ instead of own block_

Closes (https://github.com/pytorch/pytorch/issues/39810)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41560

Reviewed By: Krovatkin

Differential Revision: D22583196

Pulled By: wconstab

fbshipit-source-id: 8f543cd9ea00e116cf3e776ab168cdd9fed69632

malfet committed:
Use Python3.7 when running OSX builds/tests (#42191)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42191

Reviewed By: seemethere

Differential Revision: D22801091

Pulled By: malfet

fbshipit-source-id: b589343ef1bc6896d3d6d8d863f75aa3a102d985

venkatacrc committed:
Dequant => Swish => Quant Test case. (#41976)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41976

Dequant => Swish => Quant Test case.

(Note: this ignores all push blocking failures!)

Test Plan: test_deq_swish_quant_nnpi.py.

Reviewed By: hyuen

Differential Revision: D22718593

fbshipit-source-id: 1cee503a27e339af6d89c819007511b90bb6610c

malfet committed:
Stop building PyTorch for VS2017 (#42144)

Summary:
And since CUDA-9.2 is incompatible with VS2019, disable CUDA-9.2 for Windows as well

Fixes #{issue number}

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42144

Reviewed By: pbelevich

Differential Revision: D22794475

Pulled By: malfet

fbshipit-source-id: 24fc980e6fc75240664b9de8a4a63b1153f8d8ee

jerryzh168 committed:
[quant][graphmode] Support stack (#42187)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42187

Test Plan: Imported from OSS

Reviewed By: vkuzo

Differential Revision: D22801229

fbshipit-source-id: 7d1758c4fb1c8f742a275c3a631605f0f0d08e44

ChunliF committed:
[Shape Inference] Fix InferFC

Summary: Sometimes first dim of X in FC is BATCH_OF_FEATURE_MAX instead of BATCH. This caused an issue in f207899183 (when first dim of X is 64 but is set to 1 in inferFC). Change the check from `!= BATCH` to `== UNKNOWN`

Test Plan: unit test

Reviewed By: yinghai

Differential Revision: D22784691

fbshipit-source-id: eb66ba361d6fe75672b13edbac2fbd269a7e7a00

ipiszy committed:
Add a flag to enforce fp32 to fp16 conversion for all inputs of the onnxifi net. (#39931)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/39931

ATT.

Reviewed By: yinghai, ChunliF

Differential Revision: D21993492

fbshipit-source-id: ff386e6e9b95a783906fc1ae6a62462e6559a20b

osalpekar committed:
Fixed Skipping Logic in ProcessGroupNCCLErrors tests (#42192)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42192

This PR fixes the complicated skipping logic for ProcessGroupNCCLErrors Tests - it correctly logs the reason for skipping tests when GPUs are not available or the NCCL version is too old.

This is part of a broader effort to improve the testing of the ProcessGroup and Collectives tests.
ghstack-source-id: 108620568

Test Plan: Tested on devGPU and devvm. Tests are run correctly on GPU and skipped on CPU as expected.

Reviewed By: mrshenli

Differential Revision: D22782856

fbshipit-source-id: 6071dfdd9743f45e59295e5cee09e89c8eb299c9

osalpekar committed:
Add appropriate error messages for ProcessGroupNCCLTest (#42143)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42143

Replaces the original makeshift error messages in ProcessGroupNCCLTest
with more appropriate ones.
ghstack-source-id: 108711579

Test Plan: Ran the tests on DevGPU

Reviewed By: mrshenli

Differential Revision: D22778505

fbshipit-source-id: 27109874f0b474a74b09f588cf6e7528d2069702

pritamdamania committed:
Fix flaky NCCL error handling tests. (#42149)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42149

Some of these tests were flaky since we could kill the process in some
way without cleaning up the ProcessGroup. This resulted in issues where the
FileStore didn't clean up appropriately resulting in other processes in the
group to crash.

Fixed this by explicitly deleting the process_group before we bring a process
down forcibly.
ghstack-source-id: 108629057

Test Plan: waitforbuildbot

Reviewed By: mrshenli

Differential Revision: D22785042

fbshipit-source-id: c31d0f723badbc23b7258e322f75b57e0a1a42cf

ezyang committed:
Remove dead named_tensors_unsupported_error definitions. (#42171)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42171

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Test Plan: Imported from OSS

Reviewed By: pbelevich

Differential Revision: D22794980

Pulled By: ezyang

fbshipit-source-id: 250b6566270e19240361d758db55101d6fcb33e9

ann-ss committed:
refactor save_data as non member function (#42045)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42045

This PR changes the save_data() member functions of torch::jit::mobile::Module which was introduced in #41403 to be the non member function torch::jit::mobile::_save_parameters() (taking a mobile Module as its first argument).

In addition, this PR:
* adds a getter function _ivalue() for the mobile::Module object
* renames torch::jit::mobile::_load_mobile_data() to torch::jit::mobile_load_parameters()
* refactors the import.h header file into import.h and import_data.h

Test Plan: Imported from OSS

Reviewed By: kwanmacher, iseeyuan

Differential Revision: D22766781

Pulled By: ann-ss

fbshipit-source-id: 5cabae31927187753a958feede5e9a28d71d9e92

mattip committed:
DOC: fail to build if there are warnings (#41335)

Summary:
Merge after gh-41334 and gh-41321 (EDIT: both are merged).
Closes gh-38011

This is the last in a series of PRs to build documentation without warnings. It adds `-WT --keepgoing` to the shpinx build which will [fail the build if there are warnings](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-W), print a [trackeback on error](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-T) and [finish the build](https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-keep-going) even when there are warnings.

It should fail now, but pass once the PRs mentioned at the top are merged.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41335

Reviewed By: pbelevich

Differential Revision: D22794425

Pulled By: mruberry

fbshipit-source-id: eb2903e50759d1d4f66346ee2ceebeecfac7b094

Commit:
Enables torch.full bool and integer type inference (#41912)

Summary:
After being deprecated in 1.5 and throwing a runtime error in 1.6, we can now enable torch.full inferring its dtype when given bool and integer fill values. This PR enables that inference and updates the tests and docs to reflect this.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41912

Reviewed By: pbelevich

Differential Revision: D22790718

Pulled By: mruberry

fbshipit-source-id: 8d1eb01574b1977f00bc0696974ac38ffdd40d9e

jianyuh committed:
[pt] Add incude_last_offset option to EmbeddingBag mean and max (#42215)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42215

Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/).

More details: https://www.internalfb.com/intern/diff/D22794469/?dest_fbid=309597093427021&transaction_id=631457624153457

ghstack-source-id: 108733009

Test Plan:
```
buck test mode/dev-nosan //caffe2/test:nn -- "test_EmbeddingBag_per_sample_weights_and_new_offsets_cpu"
```

```
(base) [jianyuhuang@devbig281.ftw3.facebook.com: ~/fbsource/fbcode/caffe2/test] $ TORCH_SHOW_CPP_STACKTRACES=1 buck test mode/dev-nosan //caffe2/test:
nn -- "test_EmbeddingBag_per_sample_weights_and_new_offsets_cpu" --print-passing-details
Parsing buck files: finished in 1.2 sec
Building: finished in 5.5 sec (100%) 10130/10130 jobs, 2 updated
  Total time: 6.7 sec
More details at https://www.internalfb.com/intern/buck/build/dbdc2063-69d8-45cb-9146-308a9e8505ef
First unknown argument: --print-passing-details.
Falling back to TestPilot classic.
Trace available for this run at /tmp/testpilot.20200728-195414.1422748.log
TestPilot test runner for Facebook. See https://fburl.com/testpilot for details.
Testpilot build revision cd2638f1f47250eac058b8c36561760027d16add fbpkg f88726c8ebde4ba288e1172a348c7f46 at Mon Jul 27 18:11:43 2020 by twsvcscm from /usr/local/fbprojects/packages/testinfra.testpilot/887/t.par
Discovering tests
Running 1 test
Started new test run: https://our.intern.facebook.com/intern/testinfra/testrun/844425097242375
      ✓ caffe2/test:nn - test_EmbeddingBag_per_sample_weights_and_new_offsets_cpu (test_nn.TestNNDeviceTypeCPU) 0.162 1/1 (passed)
Test output:
> /data/users/jianyuhuang/fbsource/fbcode/buck-out/dev/gen/caffe2/test/nn#binary,link-tree/torch/_utils_internal.py:103: DeprecationWarning: This is a NOOP in python >= 3.7, its just too dangerous with how we write code at facebook. Instead we patch os.fork and multiprocessing which can raise exceptions if a deadlock would happen.
>   threadSafeForkRegisterAtFork()
> /usr/local/fbcode/platform007/lib/python3.7/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__
and __path__
>   return f(*args, **kwds)
> test_EmbeddingBag_per_sample_weights_and_new_offsets_cpu (test_nn.TestNNDeviceTypeCPU) ... Couldn't download test skip set, leaving all tests enabled...
> ok
>
> ----------------------------------------------------------------------
> Ran 1 test in 0.162s
>
> OK
Finished test run: https://our.intern.facebook.com/intern/testinfra/testrun/844425097242375
Summary (total time 5.54s):
  PASS: 1
  FAIL: 0
  SKIP: 0
  FATAL: 0
  TIMEOUT: 0
  OMIT: 0
Did _not_ run with tpx. See https://fburl.com/tpx for details.
```

Reviewed By: dzhulgakov

Differential Revision: D22801881

fbshipit-source-id: 80a624465727081bb9bf55c28419695a3d79c6e5

RockingJavaBean committed:
implement numpy-like functionality isposinf, isneginf (#41588)

Summary:
Related https://github.com/pytorch/pytorch/issues/38349

Numpy-like functionalities `isposinf` and `isneginf` are implemented.

Test-Plan:
- pytest test/test_torch.py -k "test_isposinf_isneginf"

Pull Request resolved: https://github.com/pytorch/pytorch/pull/41588

Reviewed By: ngimel

Differential Revision: D22770732

Pulled By: mruberry

fbshipit-source-id: 7448653e8fb8df6b9cd4604a4739fe18a1135578

albanD committed:
Revert D22790718: [pytorch][PR] Enables torch.full bool and integer type inference

Test Plan: revert-hammer

Differential Revision:
D22790718 (https://github.com/pytorch/pytorch/commit/6b3f335641e1874c5b0ae0aa719687ed5f42ab45)

Original commit changeset: 8d1eb01574b1

fbshipit-source-id: c321177cce129a6c83f1a7b26bd5ed94a343ac0f

facebook-github-bot committed:
Automated submodule update: FBGEMM (#42205)

Summary:
This is an automated pull request to update the first-party submodule for [pytorch/FBGEMM](https://github.com/pytorch/FBGEMM).

New submodule commit: https://github.com/pytorch/FBGEMM/commit/cad1c21404754156c24da76827e8f90fb725922f

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42205

Test Plan: Ensure that CI jobs succeed on GitHub before landing.

Reviewed By: dskhudia

Differential Revision: D22806731

Pulled By: efiks

fbshipit-source-id: 779a9f7f00645e7e65f183e2832dc79117eae5fd

Commit:
Fix type annotation for DistributedDataParallel (#42231)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/42231

Reviewed By: albanD

Differential Revision: D22816589

Pulled By: mrshenli

fbshipit-source-id: a355f7e2fa895617bf81ef681b051f074d39ab8c

Commit:
[JIT] Dont include view ops in autodiff graphs (#42027)

Summary:
View ops as outputs of differentiable subgraphs can cause incorrect differentiation. For now, do not include them in the subgraph. This was observed with our autograd tests for MultiheadAttention and nn.Transformer, which currently fail with the legacy executor. This commit fixes those test failures.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42027

Reviewed By: pbelevich

Differential Revision: D22798133

Pulled By: eellison

fbshipit-source-id: 2f6c08953317bbe013933c6faaad20100376c039

Commit:
Extending Learnable Fake Quantize module to support gradient scaling and factory (partial) construction (#41969)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41969

In this diff, the `_LearnableFakeQuantize` module is extended to provide support for gradient scaling where the gradients for both scale and zero point are multiplied by a constant `g` (in some cases, can help with quicker convergence). In addition, it is also augmented to provide a factory method via `_with_args` such that a partial constructor of the module can be built.

Test Plan:
For correctness of the fake quantizer operators, on a devvm, enter the following command:
```
buck test //caffe2/torch:quantization -- learnable_py_module
```

Reviewed By: z-a-f

Differential Revision: D22715629

fbshipit-source-id: ff8e5764f81ca7264bf9333789f57e0b0cec7a72

scintiller committed:
add tests to BinaryOpsKernel -- max/min kernel (#42198)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42198

1. add tests to max/min kernel

Test Plan:
1. Run locally to check cover the corresponding code part in BinaryOpsKernel.cpp.
2. CI

Reviewed By: malfet

Differential Revision: D22796019

fbshipit-source-id: 84c8d7df509de453c4ec3c5e38977733b0ef3457

Commit:
Operator-level Benchmark Test for Per Tensor and Per Channel Fake Quantization (#41974)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41974

In this diff, 2 new sets of benchmark tests are added to the `quantization` benchmark suite where operator-level benchmarking is conducted for the learnable Python operators, the learnable c++ kernels, and the original non-backprop c++ kernels.

Test Plan:
Inside the path `torch/benchmarks/operator_benchmark` (The root directory will be `caffe2` inside `fbcode` if working on a devvm):
- On a devvm, run the command `buck run pt:fake_quantize_learnable_test`
- On a personal laptop, run the command `python3 -m pt.fake_quantize_learnable_test`

Benchmark Results (On devGPU with 0% volatile utilization -- all GPUs are free):
Each sample has dimensions **3x256x256**;

### In **microseconds** (`1e-6` second),

|                           | Python Module | C++ Kernel | Non-backprop C++ Kernel |
|---------------------------|---------------|------------|-------------------------|
| Per Tensor CPU Forward    | 3112.666      | 3270.740   | 3596.864                |
| Per Tensor Cuda Forward   | 797.258       | 258.961    | 133.953                 |
| Per Channel CPU Forward   | 6587.693      | 6931.461   | 6352.417                |
| Per Channel Cuda Forward  | 1579.576      | 555.723    | 479.016                 |
| Per Tensor CPU Backward   | 72278.390     | 22466.648  | 12922.195               |
| Per Tensor Cuda Backward  | 6512.280      | 1546.218   | 652.942                 |
| Per Channel CPU Backward  | 74138.545     | 41212.777  | 14131.576               |
| Per Channel Cuda Backward | 6795.173      | 4321.351   | 1052.066                |

Reviewed By: z-a-f

Differential Revision: D22715683

fbshipit-source-id: 8be528b790663413cbeeabd4f68bbca00be052dd

bertmaher committed:
Environment variable for controlling type verbosity in debug output (#41906)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41906

Fixes #41770

Test Plan:
Example:
```
import torch
def bar():
    def test(a):
        return a
    x = torch.ones(10,10, device='cpu')
    print(torch.jit.trace(test, (x)).graph)
bar()
```

Bash:
```
for i in 0 1 2 3; do
  PYTORCH_JIT_TYPE_VERBOSITY=$i python test.py
done
```

Output:
```
graph(%0):
  return (%0)

graph(%0 : Float(10, 10)):
  return (%0)

graph(%0 : Float(10:10, 10:1)):
  return (%0)

graph(%0 : Float(10:10, 10:1, requires_grad=0, device=cpu)):
  return (%0)
```

Imported from OSS

Reviewed By: ZolotukhinM

Differential Revision: D22687966

fbshipit-source-id: cd395257d79a4baa35245c778a74a55d1ea2a842

xiaomengy committed:
[Caffe2] Fix spatial_bn bug for computing running_var on CPU or on CUDA without CuDNN (#42151)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42151

Previously our Caffe2 SpatialBN op impl was incorrect for computing running_var without unbias coefficent. Actually it should fail the test because the output will be different with CuDNN's output. However, our tests are too weak to find this bug. This diff fix all of them.

Test Plan: buck test mode/dev-nosan //caffe2/caffe2/python/operator_test:spatial_bn_op_test

Reviewed By: houseroad

Differential Revision: D22786127

fbshipit-source-id: db80becb67d60c44faae180c7e4257cb136a266d

Commit:
Improved coverage for unboxed->boxed kernel wrappers (#38999)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/38999

Adds boxing for inplace and outplace kernels, itemizes
remaining unsupported cases, and fails compilation when
new unsupported types are introduced in op signatures.

Test Plan: Imported from OSS

Differential Revision: D21718547

Pulled By: bhosmer

fbshipit-source-id: 03295128b21d1843e86789fb474f38411b26a8b6

Commit:
Creates spectral ops test suite (#42157)

Summary:
In preparation for creating the new torch.fft namespace and NumPy-like fft functions, as well as supporting our goal of refactoring and reducing the size of test_torch.py, this PR creates a test suite for our spectral ops.

The existing spectral op tests from test_torch.py and test_cuda.py are moved to test_spectral_ops.py and updated to run under the device generic test framework.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42157

Reviewed By: albanD

Differential Revision: D22811096

Pulled By: mruberry

fbshipit-source-id: e5c50f0016ea6bb8b093cd6df2dbcef6db9bb6b6

zhaojuanmao committed:
grad detach_ only when it has grad_fn in zero_grad call (#41283)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41283

in optimizer.zero_grad(), detach_ is useful to avoid memory leak only when grad has grad_fn, so add check to call grad.detach_ only when the grad has grad_fn in zero_grad() function
ghstack-source-id: 108702289

Test Plan: unit test

Reviewed By: mrshenli

Differential Revision: D22487315

fbshipit-source-id: 861909b15c8497f1da57f092d8963d4920c85e38

Commit:
fix typo in GradScaler docstring (#42236)

Summary:
Closes https://github.com/pytorch/pytorch/issues/42226.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/42236

Reviewed By: albanD

Differential Revision: D22817980

Pulled By: ngimel

fbshipit-source-id: 4326fe028dba1dbeed454edc4e4d4fffa56f51d6

Commit:
Add NCCL Alltoall to PT NCCL process group (#42051)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/42051

Add Alltoall and Alltoallv to PT NCCL process group using NCCL Send/Recv.

Differential Revision: D22676566

fbshipit-source-id: e7866b198eaa3212d608941405955df7dbd8f532
